### ISSUE TITLE
[REQUEST] Replace Boost.Python with PyBind11

### ISSUE THREAD

**Issue by Idclip** [AcademySoftwareFoundation/openvdb#1124]
### Looking for help with Boost.Python -> PyBind11 migration

These points represent the work required to modify python bindings, located in `openvdb/python/`, to build with `PyBind11` and remove the `boost-python` dependency. Any work on the below should be done off of the `feature/pybind` branch. This branch will be synced with `master` on releases and allows for any work to be contributed incrementally:

 - Completely replace the `boost-python` dependency for the pyopenvdb module with `PyBind11`[*]. This includes working with NumPy support.
 - Maintain API and feature parity with the existing `boost-python` bindings. The Python unit tests should ideally require no modification and pass as they are after this work has been completed.
 - Maintain the existing folder structure/file layout where possible
 - Use the existing CMake build system and work with OpenVDBs minimum version requirements, including CMake, GCC/Clang, CXX standard, Python versions >= 2.7 and `PyBind11` versions >= 2.6 [**].
 - Conform to the OpenVDB Coding standards
 - Have commits signed-off (DCO checks) and contributors should be prepared to sign a CLA.
 - Bonus: Modify the CI, where necessary, to build and test with PyBind11. See `.github/workflows/build.yml`
------------------
[*] Note that, similar to the existing build system, `PyBind11` should be expected to exist in a users environment and should *not* be incorporated into the OpenVDB repository (there are outstanding licensing questions with this approach so this should *not* be done at this stage).

[**] `PyBind11` 2.6 was released in October 2020. The latest version is 2.6.2. I think ensuring it works with 2.6.X is enough.

**Comment by chrisflesher** [comment]
Refer to issue #1047 for info on how to get started.

**Comment by chrisflesher** [comment]
I started on an attempt for this, am pretty inexperienced with pretty much all aspects though... so any collaboration is welcome.
https://github.com/chrisflesher/openvdb/tree/replace-boost-python-with-pybind11

Am using the scikit-build system and trying to get it to compile using the following commands:
```
cd openvdb/openvdb/openvdb/myopenvdb
pip3 install . 2> stderr.txt
```

**Comment by nachovizzo** [comment]
As a reference, I have been working with OpenVDB and pybind11 for 1 year now and I have already some open-source bindings:

- [vdbfusion[(https://github.com/PRBonn/vdbfusion)
- [vdb_to_numpy](https://github.com/PRBonn/vdb_to_numpy)

I don't have the time to carry on this effort right now (although I'd love to get rid of boost python!) But I hope this helps to whoever is/will be working on this issue

**Comment by diiigle** [comment]
> How should I generalize `openvdb::Vec2i` to `typename VecT` in the following example?

I would actually work with preprocessor macros here.

Refer to 
 - https://github.com/mitsuba-renderer/drjit/blob/master/src/python/bind_llvm.cpp
 - https://github.com/mitsuba-renderer/drjit/blob/master/src/python/bind.h

for some examples of complex templated binding code. They work with a combination of templated binding functions (top pf `bind.h`) and preprocessor macros (bottom of `bind.h`) to generate the binding code.

**Comment by matthewdcong** [comment]
I've been looking into pybind11 for OpenVDB. I have all of the Python bindings ported over from boost::python to pybind11 and all of the Python unit tests are passing (without modification). There are still a few issues here and there that I need to address, but it's looking promising.

**Comment by danrbailey** [comment]
> I've been looking into pybind11 for OpenVDB. I have all of the Python bindings ported over from boost::python to pybind11 and all of the Python unit tests are passing (without modification). There are still a few issues here and there that I need to address, but it's looking promising.

I'm very happy to hear that, this has been on our roadmap for a long time. Thanks for the update Matt and feel free to join one of our upcoming weekly TSC meetings if you want to discuss / present anything. :)

**Comment by Idclip** [comment]
Closing, this work has been merged into the latest 10.1 release, thanks all
